### PR TITLE
[modernize][cleanup] Prefer 'override' usage

### PR DESCRIPTION
**Pull request by hudokkow**
## Description
Reduces number of `warning: 'foobar' overrides a member function but is not marked 'override' [-Winconsistent-missing-override]` warnings from 3899 to 50 under 17.04 with clang 4.0.0

Remaining 50 warnings are `c-pluff` exclusive.

I tried to separate the changes per folder/component for easier review and/or to drop/separate unwanted commits.

libUPnP has the respective patch under patches/
gtest could also benefit from this but I chose not to touch it

Done automagically with clang-tidy.

## Motivation and Context
Cleanup.
Partial alternative to https://github.com/xbmc/xbmc/pull/12235

## How Has This Been Tested?
Build tested and brief runtime under 17.04

## Types of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [X] Cleanup

### CODE REVIEW

**Review by Paxxi — APPROVED** [review]
good thing with override, if it builds it's correct :)

**Comment by Paxxi** [comment]
note, this likely isn't complete as clang-tidy only handles files in the compile_commands file so android/windows/osx might still be missing override specifiers. It's a great start anyway and I say merge once all builds are complete.

**Comment by FernetMenta** [comment]
please submit VP changes to vpupdates or let me do this by myself (soon)

**Comment by Paxxi** [comment]
I'm fine with excluding vp as long as it gets done in a reasonable timeframe.

**Comment by FernetMenta** [comment]
as said, it will be part of vpupdates

**Comment by hudokkow** [comment]
I can drop or separate the commits and PR to vpupdates.

Just need to know which ones or what is *off limits* because https://github.com/xbmc/xbmc/pull/12212/files touches more than just cores/VideoPlayer

If it's just VP, dropping https://github.com/xbmc/xbmc/pull/12374/commits/6dce33b6413755420d8ef8584eceeadeda7d9863 should be enough.

**Comment by garbear** [comment]
I've already started this process for VP, so can you PR vpupdates separately to my branch so that it ends up in https://github.com/xbmc/xbmc/pull/12250?

**Comment by FernetMenta** [comment]
@garbear I think it is easier and faster to submit smaller chunks

**Comment by hudokkow** [comment]
Dropped VP changes.

@garbear, @FernetMenta is right. This is a high speed moving target. Fourth rebase in 10 hours...
It touches a lot of code.

VP commit is isolated in https://github.com/hudokkow/xbmc/tree/use_override_garbear but PR to https://github.com/garbear/xbmc/tree/vp-override-warnings is quite messy. Your branch needs a rebase.
Or you pick up the patch. Your choice, just let me know.

@Paxxi, correct. I can do the same for all platforms except OSX.

@Rechi, thanks for the Windows fix. :+1:

**Comment by AlwinEsch** [comment]
A question of understanding. If the destructor of a child class gets an override, is the destructor of parent still called?

**Comment by garbear** [comment]
> A question of understanding. If the destructor of a child class gets an override, is the destructor of parent still called?

**Comment by garbear** [comment]
OK I totally f*cked up your last comment and the PR in general. In my defense I'm on vacation and some day drinking definitely occurred.

> @hudokkow PR separately and I'll rebase on your commits.

Don't worry about my PR, I'll adapt to whatever you upstream. You're the hero who tackled almost 4,000 warnings.

**Comment by Rechi** [comment]
@hudokkow this needs a rebase again
@FernetMenta are all VP related changes excluded?

**Comment by hudokkow** [comment]
VP changes are in https://github.com/hudokkow/xbmc/tree/use_override_garbear. I'll PR to @garbear's repo later today.

@garbear, no problem but next time I want a beer too. Or maybe two. ;)
@rechi, thx for the windows fix

**Comment by hudokkow** [comment]
@garbear, before I rebase, any reason why you're keeping `virtual` after adding `override`?

**Comment by garbear** [comment]
> @garbear, before I rebase, any reason why you're keeping virtual after adding override?

Explicitness. It's easier to pick out a keyword from the begging of a line as opposed to the end.

**Comment by hudokkow** [comment]
OK, thx for the explanation. Unfortunately I don't think there's a way to maintain `virtual` using clang-tidy's `modernize-use-override` magic wand.

File: xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.h [hunk 1]
```diff
@@ -232,51 +232,51 @@ class CActiveAE : public IAE, public IDispResource, private CThread
...
-  virtual bool   Resume();
-  virtual bool   IsSuspended();
-  virtual void   OnSettingsChange(const std::string& setting);
-
-  virtual float GetVolume();
-  virtual void  SetVolume(const float volume);
-  virtual void  SetMute(const bool enabled);
-  virtual bool  IsMuted();
-  virtual void  SetSoundMode(const int mode);
+  void   Shutdown() override;
```
Comment (FernetMenta) [inline xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.h]: please also drop the vertical alignment. I always do this when touching things like this.

Comment (hudokkow) [inline xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.h]: Will do.

Comment (FernetMenta) [inline xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.h]: thanks, all other look fine to me

**Comment by garbear** [comment]
> OK, thx for the explanation. Unfortunately I don't think there's a way to maintain virtual using clang-tidy's modernize-use-override magic wand.

It would be nice if the code formatter could add the `virtual`. If not I demand you go through and add all 3,899 by hand 😉  j/k I'm fine without it.

Could you drop the changes to `xbmc/games/` and `/xbmc/input/joysticks/`? I'm pretty sure that all the code I maintain uses `override` properly, so it would save me some rebase headaches later on.

**Comment by hudokkow** [comment]
Requested changes by @garbear and @FernetMenta done. All green. Good to go?

**Review by FernetMenta — APPROVED** [review]

**Review by garbear — APPROVED** [review]

**Comment by StormTrooper** [comment]
I'm getting a build error on osmc nightlies with this commit. Any idea what could be causing this?

>xbmc/interfaces/legacy/wsgi/WsgiErrorStream.h:48: Error: Syntax error in input(3).
>build/build/swig/CMakeFiles/python_binding.dir/build.make:92: recipe for target 'build/build/swig/AddonModuleXbmcwsgi.i.cpp' failed
>make[2]: *** [build/build/swig/AddonModuleXbmcwsgi.i.cpp] Error 1

Line 48:  ~WsgiErrorStream() override

**Comment by hudokkow** [comment]
Thank you for the report.

Our [CI builds](http://jenkins.kodi.tv/) say we don't have a problem with this particular piece of code.

I suggest that you try to sort things out with OSMC dev(s) and if that doesn't work, post all relevant info (OS, build logs, libs used, etc.) on our [trac](https://trac.kodi.tv/) and paste the link here. If we come to the conclusion that it's an OSMC problem, it will be ignored. If it's ours we'll fix it.

This is a dev space and every message spams all the devs. Please follow the steps outlined above before posting here again.

Thx for understanding.

**Pull request merged**